feat: allow underscore prefix for private identifiers

ivanauth opened the pull request:
## Summary
- Enable underscore prefix (`_`) for definition, relation, and permission identifiers
- Establishes a convention for marking identifiers as "private" or "internal"

## Description
This PR implements the feature requested in issue #2066 by updating the identifier regex patterns from `[a-z]` to `[a-z_]` to allow identifiers to begin with an underscore.

This is useful for:
- **Synthetic permissions**: Permissions that exist only to compose other permissions
- **Internal relations**: Relations not meant to be directly referenced by application code
- **Implementation details**: Parts of your schema that may change without affecting the public API

## Changes
- Updated regex patterns in proto validation to allow underscore-prefixed identifiers
- Updated corresponding test expectations

## Example Usage
```zed
definition document {
    relation viewer: user
    relation _internal_viewer: user
    
    // Private synthetic permission
    permission _can_view = viewer + _internal_viewer
    
    // Public permission
    permission view = _can_view
}
```

## Note
End-user documentation for this feature should be added to [authzed/docs](https://github.com/authzed/docs) at `pages/spicedb/concepts/schema.mdx`.

Fixes #2066

ivanauth commented on docs/schema-language.md:
Thank you, i removed docs from this pr and created a new docs pr.

josephschorr commented:
I think if we introduce this, we should *enforce* that private identifiers cannot be accessed outside the scope of either the definition *or* the file (in composable schemas) and perhaps even enforce that they cannot be used in API calls

ivanauth commented:
@josephschorr good point - do you want me to try adding definition-scoped + API rejection by default?

ivanauth committed:
feat: allow underscore prefix for private identifiers

Update regex patterns across proto validation rules, tuple parsing,
and schema DSL to accept identifiers starting with underscore.
This enables a convention for marking definitions, relations, and
permissions as private/internal (e.g., _private_resource, _internal_owner).

Regenerated proto files from updated core.proto.

Fixes #2066

ivanauth committed:
chore: add CHANGELOG entry for underscore prefix identifiers

Mr-Pepe commented:
> I think we need to revert https://github.com/authzed/docs/pull/421 until this PR is merged.

Or merge this 😋 

The discrepancy is indeed a bit confusing.

1993alexey commented:
I was a bit confused by the updated documentation without code support, as I kept getting schema validation errors. Agree on rolling back the docs change until this MR is merged
